org/postgresql/core/CachedQuery.java

```diff
@@ -0,0 +1,60 @@
+/*-------------------------------------------------------------------------
+*
+* Copyright (c) 2015, PostgreSQL Global Development Group
+*
+*
+*-------------------------------------------------------------------------
+*/
+package org.postgresql.core;
+
+import org.postgresql.util.CanEstimateSize;
+
+/**
+ * Stores information on the parsed JDBC query.
+ * It is used to cut parsing overhead when executing the same query through {@link java.sql.Connection#prepareStatement(String)}.
+ */
+public class CachedQuery implements CanEstimateSize {
+    /**
+     * Cache key. {@link String} or {@link org.postgresql.jdbc2.CallableQueryKey}
+     */
+    public final Object key;
+    public final Query query;
+    public final boolean isFunction;
+    public final boolean outParmBeforeFunc;
+
+    private int executeCount;
```

Comment:
This class is shared between threads.  This should probably be an `AtomicInteger`?